client/src/main/java/org/apache/celeborn/client/read/LocalPartitionReader.java

```diff
@@ -0,0 +1,205 @@
+import org.apache.celeborn.common.network.protocol.TransportMessage;
+import org.apache.celeborn.common.protocol.MessageType;
+import org.apache.celeborn.common.protocol.PartitionLocation;
+import org.apache.celeborn.common.protocol.PbOpenStream;
+import org.apache.celeborn.common.protocol.PbStreamHandler;
+import org.apache.celeborn.common.util.FileChannelUtils;
+
+public class LocalPartitionReader implements PartitionReader {
+
+  private static final Logger logger = LoggerFactory.getLogger(LocalPartitionReader.class);
+  private final LinkedBlockingQueue<ByteBuf> results;
+  private final AtomicReference<IOException> exception = new AtomicReference<>();
+  private final int fetchMaxReqsInFlight;
+  private final PartitionLocation location;
+  private volatile boolean closed = false;
+  private final Thread readThread;
+
+  private final int numChunks;
+  private int returnedChunks = 0;
+  private int currentChunkIndex = 0;
+  private final FileChannel shuffleChannel;
+
+  public LocalPartitionReader(
+      CelebornConf conf,
+      String shuffleKey,
+      PartitionLocation location,
+      TransportClientFactory clientFactory,
+      int startMapIndex,
+      int endMapIndex)
+      throws IOException {
+    fetchMaxReqsInFlight = conf.clientFetchMaxReqsInFlight();
+    results = new LinkedBlockingQueue<>();
+    this.location = location;
+    PbStreamHandler streamHandle;
+    long fetchTimeoutMs = conf.clientFetchTimeoutMs();
+    try {
+      TransportClient client =
+          clientFactory.createClient(location.getHost(), location.getFetchPort(), 0);
+      TransportMessage openStreamMsg =
+          new TransportMessage(
+              MessageType.OPEN_STREAM,
+              PbOpenStream.newBuilder()
+                  .setShuffleKey(shuffleKey)
+                  .setFileName(location.getFileName())
+                  .setStartIndex(startMapIndex)
+                  .setEndIndex(endMapIndex)
+                  .setReadLocalShuffle(true)
+                  .build()
+                  .toByteArray());
+      ByteBuffer response = client.sendRpcSync(openStreamMsg.toByteBuffer(), fetchTimeoutMs);
+      streamHandle = TransportMessage.fromByteBuffer(response).getParsedPayload();
+    } catch (IOException | InterruptedException e) {
+      throw new IOException(
+          "Read shuffle file from local file failed, partition location: "
+              + location
+              + " filePath: "
+              + location.getStorageInfo().getFilePath(),
+          e);
+    }
+
+    final List<Long> chunkOffsets = new ArrayList<>(streamHandle.getChunkOffsetsList());
+    numChunks = streamHandle.getNumChunks();
+    shuffleChannel = FileChannelUtils.openReadableFileChannel(streamHandle.getFullPath());
+
+    logger.debug(
+        "Local partition reader {} index count:{} offsets:{}",
+        location.getStorageInfo().getFilePath(),
+        chunkOffsets.size(),
+        chunkOffsets);
+
+    readThread =
```

Comment:
Each partition has one thread may cause too many threads. I think it's better to pass a shared ExecutorService object to each LocalPartitionReader, and submit a reading task when `results` is polled out a chunk.